### PR TITLE
[SPARK-24112][SQL] Add `convertMetastoreTableProperty` conf

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

In Apache Spark 2.4, [SPARK-23355](https://issues.apache.org/jira/browse/SPARK-23355) fixes a bug which ignores table properties during `convertMetastore` for tables created by `STORED AS ORC/PARQUET`.

For some Parquet tables having table properties like `TBLPROPERTIES (parquet.compression 'NONE')`, it was ignored by default before Apache Spark 2.4. After upgrading cluster, Spark will write uncompressed file which is different from Apache Spark 2.3 and old.

In order to provide full backward-compatibility, this introduces additional configuration `spark.sql.hive.convertMetastoreTableProperty` to restore the previous behavior by ignoring table properties.

## How was this patch tested?

Pass the Jenkins.

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala [hunk 1]
```diff
@@ -189,10 +189,12 @@ case class RelationConversions(
...
   // Note that Spark doesn't support configurations like `hive.merge.orcfile.stripe.level`.
   private def isOrcProperty(key: String) =
-    key.startsWith("orc.") || key.contains(".orc.")
+    conf.getConf(HiveUtils.CONVERT_METASTORE_TABLE_PROPERTY) &&
+      (key.startsWith("orc.") || key.contains(".orc."))
 
   private def isParquetProperty(key: String) =
-    key.startsWith("parquet.") || key.contains(".parquet.")
+    conf.getConf(HiveUtils.CONVERT_METASTORE_TABLE_PROPERTY) &&
+      key.startsWith("parquet.") || key.contains(".parquet.")
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: Oh. Thanks!

**Comment by dongjoon-hyun** [comment]
Could you review this backward compatibility PR, @cloud-fan , @gatorsmile , @mridulm , and @HyukjinKwon ?

File: docs/sql-programming-guide.md [hunk 2]
```diff
@@ -1812,6 +1812,8 @@ working with timestamps in `pandas_udf`s to get the best performance, see
...
+  - In version 2.3 and earlier, Spark converts Parquet Hive tables by default but ignores table properties like `TBLPROPERTIES (parquet.compression 'NONE')`. This happens for ORC Hive table properties like `TBLPROPERTIES (orc.compress 'NONE')` in case of `spark.sql.hive.convertMetastoreOrc=true`, too. Since Spark 2.4, Spark supports Parquet/ORC specific table properties while converting Parquet/ORC Hive tables. As an example, `CREATE TABLE t(id int) STORED AS PARQUET TBLPROPERTIES (parquet.compression 'NONE')` would generate Snappy parquet files during insertion in Spark 2.3, and in Spark 2.4, the result would be uncompressed parquet files. To set `false` to `spark.sql.hive.convertMetastoreTableProperty` restores the previous behavior.
```
Comment (cloud-fan) [inline docs/sql-programming-guide.md]: is it really possible that a user sets the table property and wants Spark to ignore it? I'm OK with mentioning it in the migration guide, but adding a conf seems a overkill to me. cc @gatorsmile

Comment (mridulm) [reply docs/sql-programming-guide.md]: Setting a property and expecting spark to ignore it does not sound logical (spark not honoring a property is a bug IMO - which, thankfully, has been fixed in 2.4).
Having said that, I agree with you that mentioning this in migration guide might be sufficient;  we have behavior changes between versions all the time and a conf is not necessary when the change is in the right direction.

Comment (HyukjinKwon) [reply docs/sql-programming-guide.md]: +1 on ^ and @cloud-fan's.

Comment (dongjoon-hyun) [reply docs/sql-programming-guide.md]: Thank you for reviews, @cloud-fan , @mridulm , @HyukjinKwon ! I'll update like that.

**Comment by dongjoon-hyun** [comment]
Thank you all. SPARK-24112 is closed and the doc issue goes as [FOLLOWUP PR for SPARK-23355](https://github.com/apache/spark/pull/21269).

**Pull request closed without merging**